process/rating/peerHonesty/peerScore.go

```diff
@@ -0,0 +1,280 @@
+
+	return instance, nil
+}
+
+func checkParams(
+	peerHonestyConfig config.PeerHonestyConfig,
+	blackListedPkCache process.TimeCacher,
+	cache storage.Cacher,
+) error {
+	if check.IfNil(blackListedPkCache) {
+		return process.ErrNilBlackListedPkCache
+	}
+
+	if check.IfNil(cache) {
+		return process.ErrNilCacher
+	}
+
+	isDecayCoefficientOk := peerHonestyConfig.DecayCoefficient > minDecayCoefficient &&
+		peerHonestyConfig.DecayCoefficient < maxDecayCoefficient
+	if !isDecayCoefficientOk {
+		return fmt.Errorf("%w, decay coefficient should be in interval (%.2f, %.2f)",
+			process.ErrInvalidDecayCoefficient,
+			minDecayCoefficient,
+			maxDecayCoefficient,
+		)
+	}
+
+	if peerHonestyConfig.DecayUpdateIntervalInSeconds < minDecayIntervalInSeconds {
+		return fmt.Errorf("%w, decay interval in seconds should be greater or equal to %d",
+			process.ErrInvalidDecayIntervalInSeconds,
+			minDecayIntervalInSeconds,
+		)
+	}
+	if peerHonestyConfig.MinScore > 0 {
+		return fmt.Errorf("%w, MinScore value should be negative or zero",
+			process.ErrInvalidMinScore,
+		)
+	}
+	if peerHonestyConfig.MaxScore < 0 {
+		return fmt.Errorf("%w, MaxScore value should be positive or zero",
+			process.ErrInvalidMaxScore,
+		)
+	}
+	if peerHonestyConfig.UnitValue < 0 {
+		return fmt.Errorf("%w, UnitValue value should be positive or zero",
+			process.ErrInvalidUnitValue,
+		)
+	}
+
+	isBadPeerThresholdOk := peerHonestyConfig.BadPeerThreshold < 0 && peerHonestyConfig.MinScore < peerHonestyConfig.BadPeerThreshold
+	if !isBadPeerThresholdOk {
+		return fmt.Errorf("%w, BadPeerThreshold value should be in interval (MinScore, 0)",
+			process.ErrInvalidBadPeerThreshold,
+		)
+	}
+
+	return nil
+}
+
+func (pph *p2pPeerHonesty) executeDecay(ctx context.Context, handler func()) {
+	for {
+		select {
+		case <-time.After(pph.updateIntervalForDecay):
+			handler()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (pph *p2pPeerHonesty) applyDecay() {
+	pph.mut.Lock()
+	defer pph.mut.Unlock()
+
+	keys := pph.cache.Keys()
+	for _, key := range keys {
+		psObj, _ := pph.cache.Get(key)
+		if psObj == nil {
+			continue
+		}
+
+		ps, ok := psObj.(*peerScore)
+		if !ok {
+			continue
+		}
+
+		for topic, score := range ps.scores {
+			score = score * pph.decayCoefficient
+			if score < approximateZero && score > 0 {
+				score = 0
+			}
+
+			ps.scores[topic] = score
+		}
+	}
+}
+
+// ChangeScore will change the score of a public key on a provided topic
+func (pph *p2pPeerHonesty) ChangeScore(pk string, topic string, units int) {
+	pph.mut.Lock()
+	defer pph.mut.Unlock()
+
+	ps := pph.getValidPeerScore(pk)
+
+	oldValue := ps.scores[topic]
+	change := float64(units) * pph.unitValue
+
+	if change < 0 {
+		//TODO switch this to log.Trace in the future
+		log.Debug("p2pPeerHonesty.ChangeScore decrease",
+			"pk", hex.EncodeToString([]byte(pk)),
+			"current", fmt.Sprintf("%.2f", oldValue),
+			"change", fmt.Sprintf("%.2f", change),
+		)
+	}
+
+	ps.scores[topic] += change
```

Comment:
might use core.Max and core.Min here. No need to do that in checkBlackList

Reply:
I can not use core max and core min as this will always rewrite to the minimum value. Moved the checks out of checkBlackList